### PR TITLE
[bnb] Let's warn users when saving 8-bit models

### LINKED ISSUE #19480: [INT8] BLOOM series model loading back issue
### System Info

8x A100 GPUs with CUDA 11.3 driver

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Use the following script to save a INT8 quantized and try to load it back.
```
import os
import torch
import logging
import math
from transformers import AutoConfig, pipeline, AutoModelForCausalLM, AutoTokenizer

def get_max_memory_per_gpu_dict(dtype, model_name):
    """try to generate the memory map based on what we know about the model and the available hardware"""

    # figure out the memory map - the minimum per gpu required to load the model
    n_gpus = torch.cuda.device_count()

    try:
        # model_params calculation, as we don't have a model yet to do:
        # model_params = sum(dict((p.data_ptr(), p.numel()) for p in model.parameters()).values())

        config = AutoConfig.from_pretrained(model_name)
        h = config.hidden_size
        l = config.n_layer
        v = config.vocab_size
        # from https://github.com/bigscience-workshop/bigscience/tree/6917a3b5fefcf439d3485ca184b4d9f6ab605150/math#model-sizing
        model_params = l * (12 * h ** 2 + 13 * h) + v * h + 4 * h
    except:
        logging.info(f"The model {model_name} has a broken config file. Please notify the owner")
        raise

    if dtype == torch.int8:
        bytes = 1
    else:
        bytes = torch.finfo(dtype).bits / 8
    param_memory_total_in_bytes = model_params * bytes
    # add 5% since weight sizes aren't the same and some GPU may need more memory
    param_memory_per_gpu_in_bytes = int(param_memory_total_in_bytes / n_gpus * 1.10)
    logging.info(f"Estimating {param_memory_per_gpu_in_bytes / 2 ** 30:0.2f}GB per gpu for weights")

    # check the real available memory
    # load cuda kernels first and only measure the real free memory after loading (shorter by ~2GB)
    torch.ones(1).cuda()
    max_memory_per_gpu_in_bytes = torch.cuda.mem_get_info(0)[0]
    if max_memory_per_gpu_in_bytes < param_memory_per_gpu_in_bytes:
        raise ValueError(
            f"Unable to generate the memory map automatically as the needed estimated memory per gpu ({param_memory_per_gpu_in_bytes / 2 ** 30:0.2f}GB) is bigger than the available per gpu memory ({max_memory_per_gpu_in_bytes / 2 ** 30:0.2f}GB)"
        )

    max_memory_per_gpu = {i: param_memory_per_gpu_in_bytes for i in range(torch.cuda.device_count())}
    print("Max memory per gpu:", max_memory_per_gpu)
    return max_memory_per_gpu


def load_model():
    world_size = torch.cuda.device_count()
    model_name = "bigscience/bloom"
    logging.info(f"Using {world_size} gpus")
    logging.info(f"Loading model {model_name}")
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    dtype = torch.int8
    kwargs = dict(
        device_map="auto",
        max_memory=get_max_memory_per_gpu_dict(dtype, model_name),
    )
    logging.info("Using `load_in_8bit=True` to use quanitized model")
    kwargs["load_in_8bit"] = True
    model = AutoModelForCausalLM.from_pretrained(model_name, **kwargs)
    return model, tokenizer

model, tokenizer = load_model()

model.save_pretrained("int8_model/", max_shard_size="8GB")
```

When loading from the directory, having the error on:

```
RuntimeError: Only Tensors of floating point dtype can require gradients
```

During the initialization of the model.

```
import torch
import torch.distributed as dist
from transformers import pipeline, AutoModelForCausalLM, AutoTokenizer

model_name = 'int8_model/'
model = AutoModelForCausalLM.from_pretrained(model_name, torch_dtype=torch.int8)
```

### Expected behavior

The loading should pass.  Looking for a workaround on it...

### LINKED ISSUE #20247: Saving a 8-bit T5 does not work.
### System Info

Latest.

### Who can help?

@patrickvonplaten 

I am able to load a T5 model in 8-bit-format using this command:
`model = T5ForConditionalGeneration.from_pretrained(".",load_in_8bit=True, device_map="auto")`

The model works fine. Saving the model using ".save_pretrained()" also seems to work, and the file on the disk is a lot smaller. I am able to load the save model again without errors, but then the model no longer works.

Here are two sample models:
north/fine_North_large
north/fine_North_large_8bit

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

See description above

### Expected behavior

The model should not change when being saved.

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

This PR warns users that try to save 8-bit loaded models

closes #20247
closes #19480

cc @sgugger

### CODE REVIEW

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -2340,6 +2340,11 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
                 load_in_8bit=load_in_8bit,
             )
 
+        # Let's add a safety checker to warn users that `save_pretrained` cannot be used
+        # under 8_bit models
+        if load_in_8bit:
+            setattr(model, "save_pretrained", warn_uncompatible_save(model.save_pretrained))
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:2346]: Mmm, let's store an attribute `is_laoded_in_8bit` instead of wrapping the `save_pretrained` method.

**Review by sgugger — COMMENTED** [review]
Thanks for working on this. Not a fan of the actual implementation, suggested an alternative.

File: src/transformers/modeling_utils.py (hunk 1 above)
Comment (younesbelkada) [reply src/transformers/modeling_utils.py]: Thanks for the suggestion! Proposed an alternative in bc955d4

File: src/transformers/modeling_utils.py [hunk 2]
```diff
@@ -1538,6 +1538,15 @@ def save_pretrained(
...
         """
+        # Checks if the model has been loaded in 8-bit
+        if hasattr(self, "is_loaded_in_8bit"):
+            if self.is_loaded_in_8bit:
+                warnings.warn(
+                    "You are calling `save_pretrained` to a 8-bit converted model you may likely encounter unexepected"
+                    " behaviors. ",
+                    UserWarning,
+                )
+
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:1549]:
```suggestion
        # Checks if the model has been loaded in 8-bit
        if getatttr(self, "is_loaded_in_8bit", False):
            warnings.warn(
                "You are calling `save_pretrained` to a 8-bit converted model you may likely encounter unexepected"
                " behaviors. ",
                UserWarning,
            )

```

File: src/transformers/modeling_utils.py [hunk 3]
```diff
@@ -2340,6 +2349,11 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
                 load_in_8bit=load_in_8bit,
             )
 
+        # Let's add a safety checker to warn users that `save_pretrained` cannot be used
+        # under 8_bit models
+        if load_in_8bit:
+            setattr(model, "is_loaded_in_8bit", True)
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:2355]:
```suggestion
        self.is_loaded_in_8bit = load_in_8bit
```

**Review by sgugger — COMMENTED** [review]
More comment :-)

**Comment by sgugger** [comment]
Ah ah made a typo in one of my suggestions, sorry. `getattr` has less `t`s ;-)

File: src/transformers/modeling_utils.py (hunk 3 above)
Comment (younesbelkada) [reply src/transformers/modeling_utils.py]: maybe `cls.is_loaded_in_8bit` here?

**Review by sgugger — APPROVED** [review]

**Comment by peregilk** [comment]
Thanks for implementing this. This fix would absolutely have prevented the confusion explained in: [#20247](https://github.com/huggingface/transformers/issues/20247).

**Comment by younesbelkada** [comment]
Thank you very much @peregilk ! 🙏

**Pull request merged**